File: scryfall/card.go
package scryfall

import "time"

const CardObject = "card"

type Card struct {
	ArenaID         int    `json:"arena_id,omitempty"`
	ID              string `json:"id"`
	Language        string `json:"lang"`
	MTGOID          int    `json:"mtgo_id,omitempty"`
	MTGOFoilID      int    `json:"mtgo_foil_id,omitempty"`
	MultiverseIDs   []int  `json:"multiverse_ids,omitempty"`
	TCGplayerID     int    `json:"tcgplayer_id,omiteempty"`
	Object          string `json:"object"`
	OracleID        string `json:"oracle_id"`
	PrintsSearchURI string `json:"prints_search_uri"`
	RulingsURI      string `json:"rulings_uri"`
	ScryfallURI     string `json:"scryfall_uri"`
	URI             string `json:"uri"`

	AllParts       []RelatedCard `json:"all_parts,omitempty"`
	CardFaces      []CardFace    `json:"card_faces,omitempty"`
	CMC            float64       `json:"cmc"`
	Colors         Colors        `json:"colors,omitempty"`
	ColorIdentity  Colors        `json:"color_identity"`
	ColorIndicator Colors        `json:"color_indicator,omitempty"`
	EDHRECRank     int           `json:"edhrec_rank,omitempty"`
	Foil           bool          `json:"foil"`
	HandModifier   string        `json:"hand_modifier,omitempty"`
	Layout         Layout        `json:"layout"`
	Legalities     Legalities    `json:"legalities"`
	LifeModifier   string        `json:"life_modifier,omitempty"`
	Loyalty        string        `json:"loyalty,omitempty"`
	ManaCost       string        `json:"mana_cost,omitempty"`
	Name           string        `json:"name"`
	Nonfoil        bool          `json:"nonfoil"`
	OracleText     string        `json:"oracle_text,omitempty"`
	Oversized      bool          `json:"oversized"`
	Power          string        `json:"power,omitempty"`
	Reserved       bool          `json:"reserved"`
	Toughness      string        `json:"toughness,omitempty"`
	TypeLine       string        `json:"type_line"`

	Artist              string            `json:"artist,omitempty"`
	Booster             bool              `json:"booster"`
	BorderColor         string            `json:"border_color"`
	CardBackID          string            `json:"card_back_id"`
	CollectorNumber     string            `json:"collector_number"`
	Digital             bool              `json:"digital"`
	FlavorText          string            `json:"flavor_text,omitempty"`
	FrameEffects        []FrameEffect     `json:"frame_effects,omitempty"`
	Frame               Frame             `json:"frame"`
	FullArt             bool              `json:"full_art"`
	Games               []string          `json:"games"`
	HighResolutionImage bool              `json:"highres_image"`
	IllustrationID      string            `json:"illustration_id,omitempty"`
	ImageURIs           CardImageURIs     `json:"image_uris,omitempty"`
	Prices              map[string]string `json:"prices"`
	PrintedName         string            `json:"printed_name,omitempty"`
	PrintedText         string            `json:"printed_text,omitempty"`
	PrintedType         string            `json:"printed_type_line,omitempty"`
	Promo               bool              `json:"promo"`
	PromoTypes          []string          `json:"promo_types,omitempty"`
	PurchaseURIs        map[string]string `json:"purchase_uris"`
	Rarity              Rarity            `json:"rarity"`
	RelatedURIs         map[string]string `json:"related_uris"`
	ReleasedAt          string            `json:"released_at"`
	Reprint             bool              `json:"reprint"`
	ScryfallSetURI      string            `json:"scryfall_set_uri"`
	SetName             string            `json:"set_name"`
	SetSearchURI        string            `json:"set_search_uri"`
	SetType             SetType           `json:"set_type"`
	SetURI              string            `json:"set_uri"`
	SetCode             string            `json:"set"`
	StorySpotlight      bool              `json:"story_spotlight"`
	Textless            bool              `json:"textless"`
	Variation           bool              `json:"variation"`
	VariationOfID       string            `json:"variation_of,omitempty"`
	Watermark           string            `json:"watermark,omitempty"`
	Preview             CardPreview       `json:"preview"`
}

func (c *Card) ReleasedAtTime() time.Time {
	return TimestampToTime(c.ReleasedAt)
}

type CardFace struct {
	Artist         string        `json:"artist,omitempty"`
	ColorIdentity  Colors        `json:"color_indicator,omitempty"`
	Colors         Colors        `json:"colors,omitempty"`
	FlavorText     string        `json:"flavor_text,omitempty"`
	IllustrationID string        `json:"illustration_id,omitempty"`
	ImageURIs      CardImageURIs `json:"image_uris,omitempty"`
	Loyalty        string        `json:"loyalty,omitempty"`
	ManaCost       string        `json:"mana_cost,omitempty"`
	Name           string        `json:"name"`
	Object         string        `json:"object"`
	OracleText     string        `json:"oracle_text,omitempty"`
	Power          string        `json:"power,omitempty"`
	PrintedName    string        `json:"printed_name,omitempty"`
	PrintedText    string        `json:"printed_text,omitempty"`
	PrintedType    string        `json:"printed_type_line,omitempty"`
	Toughness      string        `json:"toughness,omitempty"`
	TypeLine       string        `json:"type_line"`
	Watermark      string        `json:"watermark,omitempty"`
}

type RelatedCard struct {
	ID        string `json:"id"`
	Object    string `json:"object"`
	Component string `json:"component"`
	Name      string `json:"name"`
	TypeLine  string `json:"type_line"`
	URI       string `json:"uri"`
}

type CardImageURIs struct {
	PNG        string `json:"png"`
	BorderCrop string `json:"border_crop"`
	Large      string `json:"large"`
	Normal     string `json:"normal"`
	Small      string `json:"small"`
}

type CardPreview struct {
	PreviewedAt string `json:"previewed_at,omitempty"`
	SourceURI   string `json:"source_uri,omitempty"`
	Source      string `json:"source,omitempty"`
}

File: packages/scryfall/src/types/card.ts
import {Colors} from './color';
import {ScryfallDate, URI, UUID} from './scryfall';

export enum BorderColor {
    Black = 'black',
    Borderless = 'borderless',
    Gold = 'gold',
    Silver = 'silver',
    White = 'white',
}

export type Card = {
    // Core Card Fields
    arena_id?: number;
    id: UUID;
    lang: string;
    mtgo_id?: number;
    mtgo_foil_id?: number;
    multiverse_ids?: number[];
    tcgplayer_id?: number;
    tcgplayer_etched_id?: number;
    cardmarket_id?: number;
    object: 'card';
    oracle_id: UUID;
    prints_search_uri: URI;
    rulings_uri: URI;
    scryfall_uri: URI;
    uri: URI;

    // Gameplay Fields
    all_parts?: RelatedCard[];
    card_faces?: CardFace[];
    cmc: number;
    color_identity: Colors;
    color_indicator?: Colors;
    colors?: Colors;
    edhrec_rank?: number;
    hand_modifier?: string;
    keywords: string[];
    layout: FrameLayout;
    legalities: Legalities;
    life_modifier?: string;
    loyalty?: string;
    mana_cost?: string;
    name: string;
    oracle_text?: string;
    oversized: boolean;
    penny_rank?: number;
    power?: string;
    produced_mana?: Colors;
    reserved: boolean;
    toughness?: string;
    type_line: string;

    // Print Fields
    artist?: string;
    attraction_lights?: number[];
    booster: boolean;
    border_color: BorderColor;
    card_back_id: UUID;
    collector_number: string;
    content_warning?: boolean;
    digital: boolean;
    finishes: Finish[];
    flavor_name?: string;
    flavor_text?: string;
    frame_effects?: FrameEffect[];
    frame: FrameLayout;
    full_art: boolean;
    games: Game[];
    highres_image: boolean;
    illustration_id?: UUID;
    image_status: 'missing' | 'placeholder' | 'lowres' | 'highres_scan';
    image_uris?: ImageUris;
    prices: CardPrices;
    printed_name?: string;
    printed_text?: string;
    printed_type_line?: string;
    promo: boolean;
    promo_types?: string[];
    purchase_uris: Record<string, URI>;
    rarity: Rarity;
    related_uris?: Record<string, URI>;
    released_at: ScryfallDate;
    reprint: boolean;
    scryfall_set_uri: URI;
    set_name: string;
    set_search_uri: URI;
    set_type: string;
    set_uri: URI;
    story_spotlight: boolean;
    textless: boolean;
    variation: boolean;
    variation_of?: UUID;
    security_stamp: 'oval' | 'triangle' | 'acorn' | 'circle' | 'arena' | 'heart';
    watermark: string;
    preview?: {
        previewed_at: ScryfallDate;
        source_uri: URI;
        source: string;
    };
};

type CardWithLayout<Layout, RequiredFields extends keyof Card> = Card &
    Required<Pick<Card, RequiredFields>> & {layout: Layout};

export type SplitCard = CardWithLayout<FrameLayout.Split, 'card_faces'>;
export type FlipCard = CardWithLayout<FrameLayout.Flip, 'card_faces'>;
export type DoubleFacedCard = CardWithLayout<FrameLayout.Transform | FrameLayout.ModalDfc, 'card_faces'>;
export type DoubleFacedToken = CardWithLayout<FrameLayout.DoubleFacedToken, 'card_faces'>;
export type MeldCard = CardWithLayout<FrameLayout.Meld, 'all_parts'>;

export type CardFace = {
    artist?: string;
    cmc?: number;
    color_indicator?: Colors;
    colors?: Colors;
    flavor_text?: string;
    illustration_id?: UUID;
    image_uris?: ImageUris;
    layout?: FrameLayout;
    loyalty?: string;
    mana_cost: string;
    name: string;
    object: 'card_face';
    oracle_id?: UUID;
    oracle_text?: string;
    power?: string;
    printed_name?: string;
    printed_text?: string;
    printed_type_line?: string;
    toughness?: string;
    type_line?: string;
    watermark?: string;
};

export type CardPrices = {
    usd: string;
    usd_foil: string;
    usd_etched: string;
    eur: string;
    tix: string;
};

export enum Finish {
    Etched = 'etched',
    Foil = 'foil',
    Glossy = 'glossy',
    NonFoil = 'nonfoil',
}

export enum Format {
    Standard = 'standard',
    Future = 'future',
    Historic = 'historic',
    Gladiator = 'gladiator',
    Pioneer = 'pioneer',
    Exporer = 'explorer',
    Modern = 'modern',
    Legacy = 'legacy',
    Pauper = 'pauper',
    Vintage = 'vintage',
    Penny = 'penny',
    Commander = 'commander',
    Brawl = 'brawl',
    HistoricBrawl = 'historicbrawl',
    Alchemy = 'alchemy',
    PauperCommander = 'paupercommander',
    DuelCommander = 'duel',
    OldSchool = 'oldschool',
    Premodern = 'premodern',
}

export enum FrameEffect {
    Legendary = 'legendary',
    Miracle = 'miracle',
    NyxTouched = 'nyxtouched',
    Draft = 'draft',
    Devoid = 'devoid',
    Tombstone = 'tombstone',
    ColorShifted = 'colorshifted',
    Inverted = 'inverted',
    SunMoonDfc = 'sunmoondfc',
    CompassLandDfc = 'compasslanddfc',
    OriginPlaneswalkerDfc = 'originpwdfc',
    MoonEldraziDfc = 'mooneldrazydfc',
    WaxingAndWaningMoonDfc = 'waxingandwaningmoondfc',
    Showcase = 'showcase',
    ExtendedArt = 'extendedart',
    Companion = 'companion',
    Etched = 'etched',
    Snow = 'snow',
    Lesson = 'lesson',
    ShatteredGlass = 'shatteredglass',
    ConvertDfc = 'convertdfc',
    FanDfc = 'fandfc',
    UpsideDownDfc = 'upsidedowndfc',
}

export enum FrameLayout {
    Normal = 'normal',
    Split = 'split',
    Flip = 'flip',
    Transform = 'transform',
    ModalDfc = 'modal_dfc',
    Meld = 'meld',
    Leveler = 'leveler',
    Class = 'class',
    Saga = 'saga',
    Adventure = 'adventure',
    Planar = 'planar',
    Scheem = 'scheme',
    Vanguard = 'vanguard',
    Token = 'token',
    DoubleFacedToken = 'double_faced_token',
    Emblem = 'emblem',
    Augment = 'augment',
    Host = 'host',
    ArtSeries = 'art_series',
    ReversibleCard = 'reversible_card',
}

export enum Game {
    Arena = 'arena',
    Mtgo = 'mtgo',
    Paper = 'paper',
}

export enum ImageType {
    Png = 'png',
    BorderCrop = 'border_crop',
    ArtCrop = 'art_crop',
    Large = 'large',
    Normal = 'normal',
    Small = 'small',
}

export type ImageUris = {[type in ImageType]: string};

export type Legalities = {[type in Format]: Legality};
export enum Legality {
    Banned = 'banned',
    Legal = 'legal',
    NotLegal = 'not_legal',
    Restricted = 'restricted',
}

export enum Rarity {
    Common = 'common',
    Uncommon = 'uncommon',
    Rare = 'rare',
    Special = 'special',
    MythicRare = 'mythic',
    Bonus = 'bonus',
}

export type RelatedCard = {
    id: UUID;
    object: 'related_card';
    component: 'token' | 'meld_part' | 'meld_result' | 'combo_piece';
    name: string;
    type_line: string;
    uri: URI;
};

export type Ruling = {
    object: 'ruling';
    oracle_id: UUID;
    source: 'wotc' | 'scryfall';
    published_at: ScryfallDate;
    comment: string;
};
